File: version4j/src/main/java/com/only4/algorithm/version4j/leetcode/IsPalindrome.java
package com.only4.algorithm.version4j.leetcode;

import com.only4.algorithm.version4j.extra.ListNode;

/**
 * @author zhenyu.jiang
 */
public class IsPalindrome {
    public boolean isPalindrome(ListNode head) {
        if (head == null || head.next == null) {
            return true;
        }

        // 找到链表的中点
        ListNode middle = findMiddle(head);

        // 反转后半部分链表
        ListNode secondHalf = reverseList(middle.next);
        ListNode secondHalfHead = secondHalf; // 保存用于恢复

        // 比较前半部分和反转后的后半部分
        boolean isPalindrome = true;
        ListNode firstHalf = head;

        while (secondHalf != null) {
            if (firstHalf.val != secondHalf.val) {
                isPalindrome = false;
                break;
            }
            firstHalf = firstHalf.next;
            secondHalf = secondHalf.next;
        }

        // 可选：恢复链表原始结构
        // middle.next = reverseList(secondHalfHead);

        return isPalindrome;
    }

    private ListNode findMiddle(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;

        while (fast.next != null && fast.next.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        return slow;
    }

    private ListNode reverseList(ListNode head) {
        ListNode prev = null;
        ListNode current = head;

        while (current != null) {
            ListNode next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }

        return prev;
    }
}

File: version4k/src/main/kotlin/com/only4/algorithm/version4k/leetcode/isPalindrome.kt
package com.only4.algorithm.version4k.leetcode

import com.only4.algorithm.version4k.extra.ListNode

fun isPalindrome(head: ListNode?): Boolean {
    if (head?.next == null) return true

    fun findMiddle(head: ListNode): ListNode {
        var slow = head
        var fast = head

        while (fast.next?.next != null) {
            slow = slow.next!!
            fast = fast.next!!.next!!
        }

        return slow
    }

    fun reverseList(head: ListNode?): ListNode? {
        var prev: ListNode? = null
        var current = head

        while (current != null) {
            val next = current.next
            current.next = prev
            prev = current
            current = next
        }

        return prev
    }

    // 找到链表的中点
    val middle = findMiddle(head)

    // 反转后半部分链表
    val secondHalf = reverseList(middle.next)
    val secondHalfHead = secondHalf // 保存用于恢复

    // 比较前半部分和反转后的后半部分
    var isPalindrome = true
    var firstHalf = head
    var secondHalfPtr = secondHalf

    while (secondHalfPtr != null) {
        if (firstHalf?.`val` != secondHalfPtr.`val`) {
            isPalindrome = false
            break
        }
        firstHalf = firstHalf.next
        secondHalfPtr = secondHalfPtr.next
    }

    // 可选：恢复链表原始结构
    // middle.next = reverseList(secondHalfHead)

    return isPalindrome
}

